Full Status tab improvements for VTAdmin

GuptaManan100 opened the pull request:
<!--
  Thank you for your contribution to the Vitess project.
  How to contribute: https://vitess.io/docs/contributing/
  Please first make sure there is an open Issue to discuss the feature/fix suggested in this PR.
  If this is a new feature, please mark the Issue as "RFC".
 -->

<!-- if this PR is Work in Progress please create it as a Draft Pull Request -->

## Description

<!-- A few sentences describing the overall goals of the pull request's commits. -->
<!-- If this is a bug fix and you think the fix should be backported, please write so. -->
The PR https://github.com/vitessio/vitess/pull/11438 introduced the `Full Status` tab in the vttablets in VTAdmin which shows the full replication status of the tablets. 
In this output it was noticed that the IO state and SQL state are printed as integers (because they are stored as an enum), instead of any meaningful output, so it becomes hard for the users to read this data unless they are familiar with Vitess internals on how the enum works.

This PR attempts to improve this situation by decoding these values before printing them using a replacer function in the `json.Stringify`.

Another thing that was noticed was that the semi-sync properties weren't printed if they are false. This is not desirable. We instead always want to print these settings. This PR addresses this too. It checks for these fields being explicitly set to true. If they aren't, then we set them to false.

Here are the images of the Full Status tab after the changes - 

![Screenshot 2022-10-12 at 1 46 11 PM](https://user-images.githubusercontent.com/35839558/195290607-71a0a801-f872-4f5c-9160-c67bc8bfc76d.png)

![Screenshot 2022-10-12 at 1 45 59 PM](https://user-images.githubusercontent.com/35839558/195290403-fc6fddd4-e2a5-4b73-ad79-70904d7ae9c2.png)

## Related Issue(s)

<!-- List related issues and pull requests. If this PR fixes an issue, please add it using Fixes #????  -->

## Checklist

-   [x] "Backport me!" label has been added if this change should be backported
-   [x] Tests were added or are not required
-   [x] Documentation was added or is not required

## Deployment Notes

<!-- Notes regarding deployment of the contained body of work. These should note any db migrations, etc. -->

ajm188 commented on web/vtadmin/src/components/routes/tablet/FullStatus.tsx:
```diff
@@ -1,15 +1,29 @@
+function stateReplacer(key: any, val: any) {
+    if (key === 'io_state' || key === 'sql_state') {
+        if (val === 3) {
+            return 'Replication Running';
+        } else if (val === 2) {
+            return 'Replication Connected';
+        } else if (val === 1) {
+            return 'Replication Stopped';
+        }
```
since this is already in the context of "IO Thread" or "SQL Thread" (in the ui, i mean), is "replication" maybe redundant here? what i am thinking is we could maybe just do:

```json
{
  "io_thread": "{running|connected|stopped}",
  "sql_thread": "{running|connected|stopped}"
}
```

deepthi replied:
Agree with @ajm188. Further, assuming that the source for this enum is https://github.com/vitessio/vitess/blob/dcd1d3b2e4f6d4a4f71979a6403ca889df3bb8c7/go/mysql/constants.go#L739-L744 
2 corresponds to `Connecting` not `Connected`. And we might want to complete the set and represent `0` as `Unknown`.

GuptaManan100 replied:
Fixed the strings. I didn't add the Unknown field for 0 because for the primary tablet, the IO_State and SQL_state aren't populated, which then defaults to being 0. We don't want to show Unknown state there.

ajm188 reviewed (approved):
LGTM, but a question about the presentation :)

notfelineit commented on web/vtadmin/src/components/routes/tablet/FullStatus.tsx:
```diff
@@ -1,15 +1,29 @@
     tablet: vtadmin.Tablet;
 }
 
+function stateReplacer(key: any, val: any) {
```
```suggestion
function stateReplacer(key: string, val: number) {
```

notfelineit replied:
You can add these types here! 😄

GuptaManan100 replied:
Fixed

notfelineit commented on web/vtadmin/src/components/routes/tablet/FullStatus.tsx:
```diff
@@ -30,8 +44,22 @@ const FullStatus: React.FC<Props> = ({ tablet }) => {
             </div>
         );
     }
+    if (!isNil(data) && !isNil(data.status)) {
```
Here you can simplify and do:
```suggestion
    if (data && data.status) {
```

GuptaManan100 replied:
Fixed

notfelineit commented on web/vtadmin/src/components/routes/tablet/FullStatus.tsx:
```diff
@@ -30,8 +44,22 @@ const FullStatus: React.FC<Props> = ({ tablet }) => {
         );
     }
+    if (!isNil(data) && !isNil(data.status)) {
+        if (data.status.semi_sync_primary_enabled !== true) {
+            data.status.semi_sync_primary_enabled = false;
+        }
```
Here if you are trying to set `data.status.semi_sync_primary_enabled` to a boolean, you can simplify by doing:
```suggestion
        data.status.semi_sync_primary_enabled = !!data.status.semi_sync_primary_enabled
```

GuptaManan100 replied:
Fixed

notfelineit commented:
I left a few suggestions for cleaning up the code 😄 Potentially the swaps from integer --> human readable state could be done at the API level in the future - since there could be multiple clients for this API.

deepthi commented:
> I left a few suggestions for cleaning up the code 😄 Potentially the swaps from integer --> human readable state could be done at the API level in the future - since there could be multiple clients for this API.

This would be preferable so that we don't have to keep two places in sync. However, I suspect that because we are returning a protobuf from the RPC, we don't have much of a choice.
Unless we change the protobuf itself to contain strings instead of enum values.

GuptaManan100 commented:
@notfelineit Thank-you for all the suggestions! I have fixed them all up. @deepthi is right about the reason why I implemented the changes in the typescript side instead of the API. In the typescript code, we also verify that we indeed get the result matching the fields of the proto message and their types. 
The alternate way to do to in the API would entail creating a different type called `FullStatusReadable` which has the enum changed to strings and `omitempty` removed from some of the fields that we always want printed. Then the API would return this type instead of `FullStatus`. However, we would still have the same drawback that we have now, which is to keep the two in sync. Right now it is the enum and string values we print in the UI, the alternate would be the two proto message types.

GuptaManan100 commented:
This is how the UI looks after the review comments - 
![Screenshot 2022-10-13 at 11 48 36 AM](https://user-images.githubusercontent.com/35839558/195518787-cec90253-a7a9-4628-afb1-eea993e82a49.png)

notfelineit commented:
@deepthi @GuptaManan100 that makes sense regarding the API proto! This looks good!
